documentation/High-level-overview.md

```diff
@@ -0,0 +1,166 @@
+ - Communication on task execution for a task host: Task definition, task inputs, task outputs.
+
+The transportation layer for messages are based on a .NET pipe implementattion. These pipes are specifically for windows, while the unix system implmentation has a wrap around sockets for the same functionality.
+
+The message layer has a custom serialization protocal that is specific to MSBuild.It is designed so implementing new types for new messages is easy, oinly needs to implement the `ITranslatable` interface. All message types used are known internal MSBuild types, that also have an extra field with a `string, string` disctionary to support extra user messages.
+
+## Graph build
+A graph build changes the sequence in which MSBuild processes projects. Normally a project starts execution, and when it has a dependency on another project, then that project starts to build. A graph build evaluates all projects and their relationship before starting execution of any project. This is achieved by looking at specific items in the XML (like Project Reference) to construct the dependency graph.
+
+There are a couple of different modes to run graph mode in:
+- Standard mode: Tried to work from the leaves of the dependency graph and makes sure all results are within the cache. If there is a cache miss / unexpected reference, it just schedules the missing reference for execution.
+- Strict / isolate mode: If there is a cache miss when building, the whole built is failed. This is used mostly for distributed system builds.
+
+## MSBuid Server
+In normal MSBuild execution the main process is cleared after the build ends, or after a set time limit. The MSBuild Server project aims to change that, making the entry point process and the scheduler process node separate entities. This allows processes to preserve in-memory cache and make consecutive builds faster.
+
+# Extensibilities
+MSBuild includes some extra features that are related to the build process but does not fit on the previous categories. These extensibility features are critical for the build process, but they can also be customized by third parties for their own use.
+
+## Packaging system
+MSBuild interacts with external packages in almost every build. However, the MSBuild engine does not recognize external packages as third parties, and it also does not handle external dependencies. This is done by a packaging system. The supported one being NuGet. As such, it is NuGet that is responsible for finding the packages, downloading them, and providing the MSBuild engine with a package path for the build.
+
+## Restore
+The restore operation is a built-in target within MSBuild. The main function is to walk through the project references and `packages.config` file about all the packages that need to be restored. This process is executed by NuGet, as MSBuild does not have a packaging system within the code.
+
+## Tasks
+
```

Comment:
### Custom Tasks
Lifecycle
```mermaid
graph 

T["Implement:\n extend ITask interface in .dll"] --> X["Register:\n &ltUsingTask/&gt"] --> F["Initialize:\n compile inline or load from assembly \n(TaskFactory)"] --> Tar["Use:\n &ltTarget&gt&ltMyTask/&gt&ltTarget/&gt"] --> S["Setup:\n Set input properties\n (TaskExecutionHost)"] --> E["ITask.Execute()"] --> O["Gather outputs: \n (TaskExecutionHost)"]
```

Reply:
I love this Task lifecycle, but this is more specific implementation than we are expecting from this document. It would be great if you could add this to out Custom Task documentation so we have it written down somewhere.